README
Copyright
=========

   Gbonds - a GNOME-based Savings Bond Inventory Program

   Copyright (C) 2000 Jim Evins

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

   For more details see the file COPYING.


What is Gbonds
==============

Gbonds is a program for maintaining inventories of U.S. Savings Bonds for
the GNOME desktop environment.  It allows you to track the current redemption
value and performance of your bonds.

Gbonds is similar in functionality to "Savings Bond Wizard[TM]" program
distributed by the U.S. Department of the Treasury.  Gbonds can import
inventories created by the Savings Bond Wizard.


System Requirements
===================

GTK+-2.2.x
LIBGNOME-2.2.x
LIBGNOMEUI-2.2.x
LIBGNOMEPRINT-2.2.x
LIBGNOMEPRINTUI-2.2.x
LIBXML-2.5.x
GNOME-VFS-2.2.x

All of these libraries are available as part of Gnome 2.2.

Note: to compile GBonds on RPM based systems, you will need the development
RPMs (e.g. gnome-vfs-devel) installed for each of these libraries.

Author
======

Jim Evins <evins@snaught.com>


Disclaimer
==========
GBonds is not affiliated with or endorsed by the U.S. Treasury
or the U.S. Government.


ChangeLog
2008-11-15  Jim Evins  <evins@snaught.com>

	* data/download-sites.xml:
		Search Treasury ftp site by default.
	* data/Makefile.am:
	* data/sb200506.asc:
	* data/sb200512.asc:
	* data/sb200606.asc:
	* data/sb200612.asc:
	* data/sb200706.asc:
	* data/sb200712.asc:
	* data/sb200806.asc:
	* data/sb200812.asc:
		Added historical redemption data since last release.

2008-11-15  Jim Evins  <evins@snaught.com>

	* configure.in:
		Preparation for version 2.0.3.
	* src/commands.c: (gb_cmd_help_about):
	* src/doc-sbw.c: (gb_doc_sbw_open):
	* src/doc-sbw.h:
	* src/doc-xml.c: (xml_to_doc), (xml_parse_doc), (doc_to_xml):
	* src/file.c: (open_response), (gb_file_open_real),
	(import_response), (import_real), (gb_file_save),
	(save_as_response), (gb_file_close):
	* src/gbonds.c: (main):
	* src/prefs-dialog.c: (gb_prefs_dialog_construct),
	(rdate_range_page):
	* src/prefs.c: (gb_prefs_load_settings):
	* src/print-dialog.c: (gb_print_dialog), (print_response):
	* src/print.c: (gb_print), (print_info_new), (print_font),
	(print_bold), (print_bold_end), (print_show), (print_show_right),
	(print_show_center):
	* src/recent.c: (gb_recent_get_filename):
	* src/table-model.c: (append_tables_from_dir):
	* src/ui.c: (gb_ui_init):
	* src/update.c: (add_finish_page), (read_site_list):
	* src/view.c: (gb_view_get_selected_bond_list), (gb_view_copy),
	(selection_received_cb):
	* src/window.c: (gb_window_set_doc):
		Fixed various compiler warnings if compiled with "-O1 -Wall".
	
2008-11-15  Jim Evins  <evins@snaught.com>

	* src/doc-xml.c: (gb_doc_xml_open):
		Test xmldoc -- not doc.  This typo would test doc after
		xmlParseFile() which is not initialized.
	* src/recent.c:
		Include string.h.
	* src/splash.c: (gb_splash):
		Fix 2 small memory leaks.

2004-12-23  Jim Evins  <evins@snaught.com>

	* gbonds.spec.in:
		Don't install two .desktop files.

2004-12-23  Jim Evins  <evins@snaught.com>

	* data/Makefile.am:
		Added all historical redemption data files -- they get downloaded anyway,
		the 1st time update is run.  Backed out previous @icondir@ change.
	* data/gbonds.keys.in:
		Backed out previous @icondir@ change.
	* data/sb*.asc
		Added all historical redemption data files.

2004-12-22  Jim Evins  <evins@snaught.com>

	* src/update.c: (read_remote_file_callback),
	(close_remote_file_callback), (download_done):
		Although the 2nd download still fails on FC2 (G2.6), it will not poison
		the .gbonds directory with partial downloads.

2004-12-17  Jim Evins  <evins@snaught.com>

	* data/Makefile.am:
		New rule for creating gbonds.keys: substitute @icondir@.
	* data/gbonds.keys.in:
		Removed a couple of fields,and prefixed the icon filename with @icondir@/.
	
2004-12-17  Jim Evins  <evins@snaught.com>

	* gbonds.spec.in:
		Updated.

2004-12-17  Jim Evins  <evins@snaught.com>

	* pixmaps/gbonds-application-x-gbonds.png:
	* pixmaps/gbonds.png:
	* pixmaps/gbonds_logo.png:
		Renamed image files and checked in as binary.
	* pixmaps/Makefile.am:
		Renamed image files, added app_icon class of images.
	* data/gbonds.desktop.in:
	* data/gbonds.keys.in:
		Refer to mime icon.
	* src/gbonds.c:
	* src/update.c:
		Refer to app icon.

2004-12-17  Jim Evins  <evins@snaught.com>

	* src/update.c: (download_files), (read_remote_file_callback),
	(close_remote_file_callback):
		Code cleanup and additional debug points.

2004-12-17  Jim Evins  <evins@snaught.com>

	* src/update.c: (add_downloading_page), (download_files),
	(open_remote_file_callback), (read_remote_file_callback),
	(update_progress_bar):
		Removed deprecated use of GtkProgress APIs.

2004-12-16  Jim Evins  <evins@snaught.com>

	* Makefile.am:
	* acinclude.m4:
	* configure.in:
		Screwing with build environment.
	
2004-12-16  Jim Evins  <evins@snaught.com>

	* src/debug.c: (gb_debug_init), (gb_debug):
	* src/debug.h:
		Updated debug module to the cleaner design used in the CVS version of
		glabels, which is based on the similar module in the latest version of
		gedit.
	* src/gbonds.c: (main):
		Updated initialization of debug module.

2004-12-16  Jim Evins  <evins@snaught.com>

	* src/doc-sbw.c: (gb_doc_sbw_open):
		Fixed to match prototype.

2004-12-16  Jim Evins  <evins@snaught.com>

	* src/doc-xml.c: (xml_parse_doc):
		Fixed to match prototype.

2004-12-16  Jim Evins  <evins@snaught.com>

	* src/file.c: (gb_file_open), (open_response), (gb_file_import),
	(import_response), (gb_file_save_as), (save_as_response):
		Imported the wrong version of this file previously.  This one actually
		has all the glabels artifacts removed.

2004-12-16  Jim Evins  <evins@snaught.com>

	* src/doc.c: (gb_doc_get_type):
	* src/hig.c: (gb_hig_alert_get_type), (gb_hig_dialog_get_type),
	(gb_hig_category_get_type), (gb_hig_vbox_get_type),
	(gb_hig_hbox_get_type):
	* src/hig.h:
	* src/view.c: (gb_view_get_type):
	* src/view.h:
	* src/window.c: (gb_window_get_type):
	* src/window.h:
	* src/prefs-dialog.c: (gb_prefs_dialog_get_type):
	* src/table-model.c: (gb_table_model_get_type):
		Cleanup of get_type functions: unify style, and make sure they
		return GType.
		
	* src/Makefile.am:
	* src/file.c: (gl_file_open), (open_response), (gb_file_open_real),
	(import_real), (gl_file_save_as), (save_as_response):
		Use new file chooser, if available.

	* src/ui.c: (gb_ui_init), (gb_ui_unref):
	* src/ui.h:
	* src/window.c: (gb_window_destroy):
	* src/window.h:
		Added book-keeping to unreference recent-view at window destruction.

2004-12-16  Jim Evins  <evins@snaught.com>

	* data/Makefile.am:
		Added 2004 redemption files.  Added new freedesktop.org mime database
		stuff.
	* data/gbonds.desktop.in:
		Brought up-to-date.
	* data/gbonds.xml.in:
		New mime data database file.
	* data/sb200406.asc:
	* data/sb200412.asc:
		2004 redemption files.

2004-12-15  Jim Evins  <evins@snaught.com>

	* configure.in:
	* src/recent-files/Makefile.am:
	* src/recent-files/egg-recent-item.c:
	* src/recent-files/egg-recent-item.h:
	* src/recent-files/egg-recent-model.c:
	* src/recent-files/egg-recent-util.c:

src/ui.c
/*
 *  (GBONDS) GNOME based Savings Bond Inventory Program
 *
 *  ui.c:  gbonds ui module
 *
 *  Copyright (C) 2001-2003  Jim Evins <evins@snaught.com>.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
 */
#include <config.h>

#include "recent-files/egg-recent-view.h"
#include "recent-files/egg-recent-view-bonobo.h"
#include <gconf/gconf-client.h>

#include "ui.h"
#include "commands.h"
#include "recent.h" 
#include "file.h"
#include "prefs.h"

#include "debug.h"

/*==========================================================================*/
/* Private macros and constants.                                            */
/*==========================================================================*/
#define GBONDS_UI_XML GBONDS_UI_DIR "gbonds-ui.xml"

/*==========================================================================*/
/* Private types.                                                           */
/*==========================================================================*/


/*==========================================================================*/
/* Private globals                                                          */
/*==========================================================================*/

static BonoboUIVerb gb_ui_verbs [] = {
	BONOBO_UI_VERB ("FileNew",               gb_cmd_file_new),
	BONOBO_UI_VERB ("FileOpen",              gb_cmd_file_open),
	BONOBO_UI_VERB ("FileImport",            gb_cmd_file_import),
	BONOBO_UI_VERB ("FileSave",              gb_cmd_file_save),
	BONOBO_UI_VERB ("FileSaveAs",            gb_cmd_file_save_as),
	BONOBO_UI_VERB ("FilePrint",             gb_cmd_file_print),
	BONOBO_UI_VERB ("FileClose",             gb_cmd_file_close),
	BONOBO_UI_VERB ("FileExit",              gb_cmd_file_exit),
	BONOBO_UI_VERB ("EditCut",               gb_cmd_edit_cut),
	BONOBO_UI_VERB ("EditCopy",              gb_cmd_edit_copy),
	BONOBO_UI_VERB ("EditPaste",             gb_cmd_edit_paste),
	BONOBO_UI_VERB ("EditAdd",               gb_cmd_edit_add),
	BONOBO_UI_VERB ("EditDelete",            gb_cmd_edit_delete),
	BONOBO_UI_VERB ("EditTitle",             gb_cmd_edit_title),
	BONOBO_UI_VERB ("EditSelectAll",         gb_cmd_edit_select_all),
	BONOBO_UI_VERB ("EditUnSelectAll",       gb_cmd_edit_unselect_all),
	BONOBO_UI_VERB ("SettingsPreferences",   gb_cmd_settings_preferences),
	BONOBO_UI_VERB ("SettingsUpdate",        gb_cmd_settings_update),
	BONOBO_UI_VERB ("HelpContents",          gb_cmd_help_contents),
	BONOBO_UI_VERB ("About",                 gb_cmd_help_about),

	BONOBO_UI_VERB_END
};

static gchar* doc_verbs [] = {
	"/commands/FileSave",
	"/commands/FileSaveAs",
	"/commands/FilePrint",
	"/commands/FilePrintPreview",
	"/commands/FileClose",
	"/commands/FileCloseAll",
	"/commands/EditUndo",
	"/commands/EditRedo",
	"/commands/EditCut",
	"/commands/EditCopy",
	"/commands/EditPaste",
	"/commands/EditAdd",
	"/commands/EditTitle",
	"/commands/EditDelete",
	"/commands/EditSelectAll",
	"/commands/EditUnSelectAll",

	NULL
};

static gchar* doc_modified_verbs [] = {
	"/commands/FileSave",

	NULL
};

static gchar* selection_verbs [] = {
	"/commands/EditCut",
	"/commands/EditCopy",
	"/commands/EditDelete",
	"/commands/EditUnSelectAll",

	NULL
};


/*==========================================================================*/
/* Local function prototypes                                                */
/*==========================================================================*/

static void view_menu_item_toggled_cb     (BonoboUIComponent           *ui_component,
					   const char                  *path,
					   Bonobo_UIComponent_EventType type,
					   const char                  *state,
					   BonoboWindow                *win);

static void set_app_main_toolbar_style 	  (BonoboUIComponent           *ui_component);

static void set_view_style                (BonoboUIComponent           *ui_component);

static void set_verb_sensitive            (BonoboUIComponent           *ui_component,
					   gchar                       *cname,
					   gboolean                     sensitive);

static void set_verb_list_sensitive       (BonoboUIComponent           *ui_component,
					   gchar                      **vlist,
					   gboolean                     sensitive);

static void set_verb_state                (BonoboUIComponent           *ui_component,
					   gchar                       *cname,
					   gboolean                     state);



/*****************************************************************************/
/* Initialize UI component for given window.                                 */
/*****************************************************************************/
void
gb_ui_init (BonoboUIComponent *ui_component,
	    BonoboWindow      *win)
{
        EggRecentView    *recent_view;
	EggRecentModel   *recent_model;
	BonoboControl    *control;

	gb_debug (DEBUG_UI, "START");

	gb_debug (DEBUG_UI, "window = %p", win);

	g_return_if_fail (ui_component != NULL);

	bonobo_ui_engine_config_set_path (bonobo_window_get_ui_engine (win),
					  "/gbonds/UIConfig/kvps");
	gb_debug (DEBUG_UI, "Path set");

	bonobo_ui_util_set_ui (ui_component,
			       "", GBONDS_UI_XML, "gbonds", NULL);
	gb_debug (DEBUG_UI, "UI set");

	bonobo_ui_component_add_verb_list_with_data(ui_component,
						    gb_ui_verbs, win);
	gb_debug (DEBUG_UI, "verb list added");

	/* Set the toolbar style according to prefs */
	set_app_main_toolbar_style (ui_component);
		
	/* Add listener for the view menu */
	bonobo_ui_component_add_listener (ui_component, "ViewMainToolbar", 
			(BonoboUIListenerFn)view_menu_item_toggled_cb, 
			(gpointer)win);

	bonobo_ui_component_add_listener (ui_component, "MainToolbarSystem", 
			(BonoboUIListenerFn)view_menu_item_toggled_cb, 
			(gpointer)win);
	bonobo_ui_component_add_listener (ui_component, "MainToolbarIcon", 
			(BonoboUIListenerFn)view_menu_item_toggled_cb, 
			(gpointer)win);
	bonobo_ui_component_add_listener (ui_component, "MainToolbarIconText", 
			(BonoboUIListenerFn)view_menu_item_toggled_cb, 
			(gpointer)win);
	bonobo_ui_component_add_listener (ui_component, "MainToolbarTooltips", 
			(BonoboUIListenerFn)view_menu_item_toggled_cb, 
			(gpointer)win);

	set_verb_list_sensitive (ui_component, doc_verbs, FALSE);

	/* add a GeditRecentView object */
        recent_model = gb_recent_get_model ();
        recent_view  =
		EGG_RECENT_VIEW (egg_recent_view_bonobo_new (ui_component,
							     "/menu/File/Recents"));
        egg_recent_view_set_model (recent_view, recent_model);
        
        g_signal_connect (G_OBJECT (recent_view), "activate",
                          G_CALLBACK (gb_file_open_recent), win);

	/* Squirrel away a copy to be unreferenced in gl_ui_unref() */
	g_object_set_data (G_OBJECT (ui_component), "recent-view", recent_view);
	gb_debug (DEBUG_UI, "END");
}

/*****************************************************************************/
/* Unref wrapper.                                                            */
/*****************************************************************************/
void
gb_ui_unref (BonoboUIComponent *ui_component)
{
	EggRecentView *recent_view;

	/* Pull out recent view to unreference. */
	recent_view = g_object_get_data (G_OBJECT(ui_component), "recent-view");
	if (recent_view) {
		g_object_unref (recent_view);
	}

	bonobo_object_unref(ui_component);
}

/*****************************************************************************/
/* Update all verbs of given UI component.                                   */
/*****************************************************************************/
void
gb_ui_update_all (BonoboUIComponent *ui_component,
		  gbView            *view)
{
	gbDoc *doc;

	gb_debug (DEBUG_UI, "START");

	bonobo_ui_component_freeze (ui_component, NULL);

	set_verb_list_sensitive (ui_component, doc_verbs, TRUE);

	doc = view->doc;
	g_return_if_fail (doc != NULL);

	set_verb_sensitive (ui_component, "/commands/EditUndo",
			    gb_doc_can_undo (doc));
	set_verb_sensitive (ui_component, "/commands/EditRedo",
			    gb_doc_can_redo (doc));

	set_verb_list_sensitive (ui_component, 
				 doc_modified_verbs,
				 gb_doc_is_modified (doc));

	set_verb_list_sensitive (ui_component,
				 selection_verbs,
				 !gb_view_is_selection_empty (view));

	bonobo_ui_component_thaw (ui_component, NULL);

	gb_debug (DEBUG_UI, "END");
}

/*****************************************************************************/
/* Update all verbs of given UI component to "no document" state.            */
/*****************************************************************************/
void
gb_ui_update_nodoc (BonoboUIComponent *ui_component)
{
	gb_debug (DEBUG_UI, "START");

	bonobo_ui_component_freeze (ui_component, NULL);
	
	set_verb_list_sensitive (ui_component, doc_verbs, FALSE);

	bonobo_ui_component_thaw (ui_component, NULL);

	gb_debug (DEBUG_UI, "END");
}

/*****************************************************************************/
/* Update doc modified verbs of given UI component.                        */
/*****************************************************************************/
void
gb_ui_update_modified_verbs (BonoboUIComponent *ui_component,
			     gbDoc             *doc)
{
	gb_debug (DEBUG_UI, "START");

	bonobo_ui_component_freeze (ui_component, NULL);

	set_verb_list_sensitive (ui_component, 
				 doc_modified_verbs,
				 gb_doc_is_modified (doc));

	bonobo_ui_component_thaw (ui_component, NULL);

	gb_debug (DEBUG_UI, "END");
}

/*****************************************************************************/
/* Update verbs associated with selection state of given UI component.       */
/*****************************************************************************/
void
gb_ui_update_selection_verbs (BonoboUIComponent *ui_component,
			      gbView            *view)
{
	gb_debug (DEBUG_UI, "START");

	bonobo_ui_component_freeze (ui_component, NULL);

	set_verb_list_sensitive (ui_component,
				 selection_verbs,
				 !gb_view_is_selection_empty (view));

	bonobo_ui_component_thaw (ui_component, NULL);

	gb_debug (DEBUG_UI, "END");
}

/*****************************************************************************/
/* Update undo/redo verbs of given UI component.                             */
/*****************************************************************************/
void
gb_ui_update_undo_redo_verbs (BonoboUIComponent *ui_component,
			      gbDoc             *doc)
{
	gb_debug (DEBUG_UI, "START");

	bonobo_ui_component_freeze (ui_component, NULL);

	set_verb_sensitive (ui_component,
			    "/commands/EditUndo", gb_doc_can_undo (doc));

	set_verb_sensitive (ui_component,
			    "/commands/EditRedo", gb_doc_can_redo (doc));

	bonobo_ui_component_thaw (ui_component, NULL);

	gb_debug (DEBUG_UI, "END");
}

/*---------------------------------------------------------------------------*/
/* PRIVATE.  View menu item toggled callback.                                */
/*---------------------------------------------------------------------------*/
static void
view_menu_item_toggled_cb (BonoboUIComponent           *ui_component,
			   const char                  *path,
			   Bonobo_UIComponent_EventType type,
			   const char                  *state,
			   BonoboWindow                *win)
{
	gboolean s;

	gb_debug (DEBUG_UI, "");

	s = (strcmp (state, "1") == 0);

	if (strcmp (path, "ViewMainToolbar") == 0)
	{
		gb_prefs->main_toolbar_visible = s;
		set_app_main_toolbar_style (ui_component);
		gb_prefs_save_settings ();

		return;
	}

	if (s && (strcmp (path, "MainToolbarSystem") == 0))
	{		
		gb_prefs->main_toolbar_buttons_style = GB_TOOLBAR_SYSTEM;
		set_app_main_toolbar_style (ui_component);
		gb_prefs_save_settings ();

		return;
	}

	if (s && (strcmp (path, "MainToolbarIcon") == 0))
	{		
		gb_prefs->main_toolbar_buttons_style = GB_TOOLBAR_ICONS;
		set_app_main_toolbar_style (ui_component);
		gb_prefs_save_settings ();

		return;
	}

	if (s && (strcmp (path, "MainToolbarIconText") == 0))
	{		
		gb_prefs->main_toolbar_buttons_style = GB_TOOLBAR_ICONS_AND_TEXT;
		set_app_main_toolbar_style (ui_component);
		gb_prefs_save_settings ();

		return;
	}

	if (strcmp (path, "MainToolbarTooltips") == 0)
	{
		gb_prefs->main_toolbar_view_tooltips = s;
		set_app_main_toolbar_style (ui_component);
		gb_prefs_save_settings ();

		return;
	}

}

/*---------------------------------------------------------------------------*/
/* PRIVATE.  Set main toolbar style.                                         */
/*---------------------------------------------------------------------------*/
static void
set_app_main_toolbar_style (BonoboUIComponent *ui_component)
{
	GConfClient *client;
	gboolean     labels;

	gb_debug (DEBUG_UI, "START");

	g_return_if_fail (BONOBO_IS_UI_COMPONENT (ui_component));
			
	bonobo_ui_component_freeze (ui_component, NULL);

	/* Updated view menu */
	set_verb_state (ui_component, 
			"/commands/ViewMainToolbar",
			gb_prefs->main_toolbar_visible);

	set_verb_sensitive (ui_component, 
			    "/commands/MainToolbarSystem",
			    gb_prefs->main_toolbar_visible);
	set_verb_sensitive (ui_component, 
			    "/commands/MainToolbarIcon",
			    gb_prefs->main_toolbar_visible);
	set_verb_sensitive (ui_component, 
			    "/commands/MainToolbarIconText",
			    gb_prefs->main_toolbar_visible);
	set_verb_sensitive (ui_component, 
			    "/commands/MainToolbarTooltips",
			    gb_prefs->main_toolbar_visible);

	set_verb_state (ui_component, 
			"/commands/MainToolbarSystem",
			gb_prefs->main_toolbar_buttons_style == GB_TOOLBAR_SYSTEM);

	set_verb_state (ui_component, 
			"/commands/MainToolbarIcon",
			gb_prefs->main_toolbar_buttons_style == GB_TOOLBAR_ICONS);

	set_verb_state (ui_component, 
			"/commands/MainToolbarIconText",
			gb_prefs->main_toolbar_buttons_style == GB_TOOLBAR_ICONS_AND_TEXT);

	set_verb_state (ui_component, 
			"/commands/MainToolbarTooltips",
			gb_prefs->main_toolbar_view_tooltips);

	
	/* Actually update main_toolbar style */
	bonobo_ui_component_set_prop (
		ui_component, "/MainToolbar",
		"tips", gb_prefs->main_toolbar_view_tooltips ? "1" : "0",
		NULL);
	
	switch (gb_prefs->main_toolbar_buttons_style)
	{
		case GB_TOOLBAR_SYSTEM:
						
			client = gconf_client_get_default ();
			if (client == NULL) 
				goto error;

			labels = gconf_client_get_bool (client, 
					"/desktop/gnome/interface/toolbar-labels", NULL);

			g_object_unref (G_OBJECT (client));
			
			if (labels)
			{			
				bonobo_ui_component_set_prop (
					ui_component, "/MainToolbar", "look", "both", NULL);
			
			}
			else
			{
				bonobo_ui_component_set_prop (
					ui_component, "/MainToolbar", "look", "icons", NULL);
			}
	
			break;
			
		case GB_TOOLBAR_ICONS:
			bonobo_ui_component_set_prop (
				ui_component, "/MainToolbar", "look", "icon", NULL);
			
			break;
			
		case GB_TOOLBAR_ICONS_AND_TEXT:
			bonobo_ui_component_set_prop (
				ui_component, "/MainToolbar", "look", "both", NULL);
			
			break;
		default:
			goto error;
			break;
	}
	
	bonobo_ui_component_set_prop (
			ui_component, "/MainToolbar",
			"hidden", gb_prefs->main_toolbar_visible ? "0":"1", NULL);

 error:
	bonobo_ui_component_thaw (ui_component, NULL);

	gb_debug (DEBUG_UI, "END");
}


/*---------------------------------------------------------------------------*/
/* Set sensitivity of verb.                                                  */
/*---------------------------------------------------------------------------*/
static void
set_verb_sensitive (BonoboUIComponent  *ui_component,
		    gchar              *cname,
		    gboolean            sensitive)
{
	gb_debug (DEBUG_UI, "START");

	g_return_if_fail (cname != NULL);
	g_return_if_fail (BONOBO_IS_UI_COMPONENT (ui_component));

	bonobo_ui_component_set_prop (ui_component,
				      cname,
				      "sensitive",
				      sensitive ? "1" : "0",
				      NULL);

	gb_debug (DEBUG_UI, "END");
}

/*---------------------------------------------------------------------------*/
/* Set sensitivity of a list of verbs.                                       */
/*---------------------------------------------------------------------------*/
static void
set_verb_list_sensitive (BonoboUIComponent   *ui_component,
			 gchar              **vlist,
			 gboolean             sensitive)
{
	gb_debug (DEBUG_UI, "START");

	g_return_if_fail (vlist != NULL);
	g_return_if_fail (BONOBO_IS_UI_COMPONENT (ui_component));

	for ( ; *vlist; ++vlist)
	{
		bonobo_ui_component_set_prop (ui_component,
					      *vlist,
					      "sensitive",
					      sensitive ? "1" : "0",
					      NULL);
	}

	gb_debug (DEBUG_UI, "END");
}

/*---------------------------------------------------------------------------*/
/* Set state of a verb.                                                      */
/*---------------------------------------------------------------------------*/
static void
set_verb_state (BonoboUIComponent   *ui_component,
		gchar               *cname,
		gboolean             state)
{
	gb_debug (DEBUG_UI, "START");

	g_return_if_fail (cname != NULL);
	g_return_if_fail (BONOBO_IS_UI_COMPONENT (ui_component));

	bonobo_ui_component_set_prop (ui_component,
				      cname,
				      "state",
				      state ? "1" : "0",
				      NULL);

	gb_debug (DEBUG_UI, "END");
}

